Pq graph for optimization and python/c++ translation

Marclie opened the pull request:
Here are the changes that introduce the pq_graph module into pdaggerq. This is a separate module that will take in equations from the pq_helper object. It uses single expression optimization and subexpression elimination by using a directed acyclic graph representation of the contractions. The graph can be visualized using graphiz for a visualization of the equations.

Marclie committed:
External line positions were not explicitly checked.

Marclie committed:
External line positions were not explicitly checked. Updated graph generation to use internal connection map. Removing pack=true again.

Marclie committed:
Revert change to graph generation.

Marclie committed:
Modified several files that alter the way the graphs are being generated.

Marclie committed:
Jusat discovered this whole time that there was a convenient flag to switch the direction of the arrow without having to swap the nodes. This made the printing much easier.

Marclie committed:
Refactor and rename 'tabuilder' to 'pq_graph'

This commit renamea the 'tabuilder' module to 'pq_graph'. The name 'pq graph' is more consistent with the logic of the algorithm. All related files, classes, and references are updated to reflect this change. In addition, the source files of this module are moved outside of the 'pdaggerq' folder for clearer separation of the module implementations. Some redundant comments and TODOs are also cleaned up for better readability. The structure and functionality of the code remain unchanged.

Marclie committed:
Refactor 'using namespace' directive for better coding practice

Removed 'using namespace std' directives from multiple files and replaced them with explicit std::usings. This is done to eliminate potential conflicts with outside libraries. It adds the advantage of allowing more explicit control over the functions and types being used, avoiding potential naming collisions and confusion as the project gets larger.

Marclie committed:
Merged subfolders 'collections' and 'misc' into the parent 'include' directory. There were too few files in those directories to justify its existence. Refactored the relevant include directives in other files to reflect this change.

Marclie committed:
Merge branch 'edeprince3:master' into pq_graph

Marclie committed:
initial commit of README.md for `pq graph` module

Marclie committed:
Refactor file paths in CMakeLists.txt

The paths for scaling_map.hpp and linkage_set.hpp have been updated to match their new locations. These files were moved in a previous commit but the changes in `CMakeLists.txt` was not pushed.

Marclie committed:
Add Apache license header to all source and header files in pq_graph module

Marclie committed:
Add README with example usage and logo (pending approval)

In pq_graph.cc, instead of setting the "make_einsum" flag to use python, the new option is "output" that takes either "python" or "c++" as arguments.

Marclie committed:
Fixes the image path

Marclie committed:
syntax for loading a binary was incorrect: the function does not return a pq_helper instance. It's not a bad idea, but can be done later.

Marclie committed:
Extracted functions related to generating a DOT file for graph visualization into a single source file for clarity. the `write_dot` functions in `PQGraph`, `Equation` and `Linkage` were moved to `dot_generator.cc`.

Also surpressed deprecated function warnings from pybind11.

Marclie committed:
Modified graph generation to have a clearer layout and also fixed the logic of the edge connectivity to represent traditional feynman diagrams.

Modified the method of contructing the flop and memory scalings in the Linkage class so that it is computed as the internal and external lines are computed.

Marclie committed:
Replace uint8_t with uint_fast8_t to optimize performance

This guarantees the fastest unsigned int type with at least 8 bits based on the compiler architecture.

Marclie committed:
Iterator functionality has been outlined but not fully implemented. The functionality is intended to replace generating all the vertices in a vector to perform operations. When working, the iterator will allow us to break out of any lookups before visiting all vertices involved in the linkage.

Marclie committed:
Add method to check if vertex is temporary and improve logic.

Added method `is_temp` to check if a vertex is a temporary object. Made improvements to graph making group together nodes that are contracted and their external lines. The `Term::is_compatible` function and the function in the substitute functions incorrectly filter out invalid contractions. these now always return true, but the logic needs to be fixed to get a better speedups.

Marclie committed:
Refactor and clean-up of Linkage class functionality.

Removed code relating building an iterator for the linkage class that was a bust. This should be implemented in principle, but for now removing it reduces clutter.

Refactored the Term class to test the new scaling of a term without affecting the original term. This prevents having to construct a new Term object to find the modified scaling. Made further changes to representation of the graph structure of equations in dot_generator.cc. Other changes are temporary fixes while logic is being debugged.

Marclie committed:
Refactored many functions in pq_graph.cc to be in consolidate.cc. Removed support for the old method of building the pq_graph object (formerly called tabuilder) by using lists of strings. Now it depends solely on the pq_helper object with an optional string for the name of the equation. Refactored several functions in Equation, Term, and Vertex to accomplish this.

Marclie committed:
Old building method using lists of strings has been removed and now only relies on the pq_helper object. Fixed issue with handling scalars in the linkage class. Adjusted workflow for constructing lines.

Marclie committed:
Old building method using lists of strings has been removed and now only relies on the pq_helper object. Fixed issue with handling scalars in the linkage class. Adjusted workflow for constructing lines.

Marclie committed:
Issues with `is_compatible` and `valid_ops` has been resolved. the function `Vertex::equivalent` and equality operator for vertices incorrectly only compared the base name of the vertex. for regular Vertex instances, this is fine since the lines are guaranteed to be the same. However, this is not so straightforward of an assumption with the linkage class. After changing these functions to directly compare the names, the functionality works as it used to. This should be remedied in the future by overloading the `Linkage::equivalent` function, but this works for now.

Marclie committed:
Removed linear searches over the internal lines to find matching indicies with a slightly more improved binary search. It was pretty inefficient, especially considering all the work I did to get the internal lines using a map. This still isn't the best approach, but it works well for now.

Marclie committed:
Refactored and simplify term permutation functionality

Methods for creating string representations of the different permutations (make_perm_string, p1_permute_string, p2_permute_string, etc.) have been replaced with a generic permute method generates a Term vector. Some unused functions and data members have been removed.

Marclie committed:
Modified linkage class to preserve the order of external lines for accurate substitutions. Relevant changes were made within the Line and Vertex class to accommodate the new workflow. Fixed some printing issues with permutations in the Term class. Made some modifications for generating linkages. Changed how the tmp containers are sorted by adding a check to make assignment terms appear first.

Marclie committed:
Substitution procedure has changed to expand permutations and allow them to be merged. This helps prevent so many permutations from being applied at once and temps can be applied to permutations for different equations at once. Also fixed some formatting of the linkage.cc class to preserve the order of indices in a contraction. Improved functionality of generating linkages so that not every term needs to generate all of its linkages and can be reused every substitution to find nested temps earlier on. Also modified the sort function for the temp equations to properly order nested temps.

Marclie committed:
Improved ordering of intermediates such that nested intermediates show up immediately after each other. This prevents storing the original intermediate in memory for too long. Also added logic to remove intermediates that only show up once after it is nested into another intermediate.

Cleaned up some of the options and their documentation. Added the ability to specify the max size of an intermediate to prevent storage of too large containers if the user desires.

Fixed functionality of `Linkage::to_vector` when requesting to not fully expand the linkage chain. To do this, I added another mutable container for lazy evaluation of the nonfully expanded term.

Refactored the Line class to have more descriptive names for the occupied/virtual status and the block type

Fixed some issues with rank 1 containers for sigma build being considered as a scalar. Also added functionality to remove the sigma index during optimization. Next commit will separate tmps with a sigma index separately.

Marclie committed:
Temporarily disabled remove_redundant_tmps due to issues. see TODO.

Marclie committed:
Added comment regarding how intermediates merge their coefficients in `consolidate.cc`.

Marclie committed:
Add term substitute operations file

Added a new file term_substitute.cc that handles all substitute-related operations including substitution, dot product making, and permutations, which were previously part of term.cc file. This aims to improve code readability and separations of concerns by shifting the related operations to a separate dedicated file.

Marclie committed:
Optimize code readability in 'pq_graph' module and enhance functionality for eom build

adjusted function 'collect_scaling' to accept an additional parameter 'include_reuse'. This allows control over whether to include 'reuse_tmps' equation in equations scaling. Fixed some issues with formatting for EOM build

Marclie committed:
"Change substitution condition for temporary equations

This commit changes the condition to allow substitution in temporary equations. Previously, the equations were set to not allow substitution by default, which led to ineffective code execution for EOM builds. Now the equations are by default enabled to allow substitutions. This increases the flexibility and efficiency of code execution, especially for EOM builds.

Also, terms_[0] has been replaced with terms_.front() for the initial bottleneck flop and memory scaling, improving the code readability.

The commit wholly aims at enhancing the functionality for EOM build and optimizing the code readability in 'pq_graph' module."

Marclie committed:
"Add conditional formats and modify temporary equation substitutions

This commit introduces major changes in the 'substitute' function. Now it accepts a boolean flag 'format_sigma'. When enabled, the formatted scalars, and their corresponding terms and linkages, are only processed if 'format_sigma' is true, it modifies the behavior of temporary equations.

The "substitute" function now adjusts the type of temporary variable to substitute based on the 'format_sigma', and changes the behavior for merging terms. It also affects the behavior for adding scalar terms to scalar equations.

The commit introduced a new function 'sort_tmps' that sorts the order of temporary variables in terms. This helps manage interplay between the declaration and usage of intermediate variables (tmps).

Overall, the changes enhance the flexibility for equation optimization, thereby improving performance in the code execution."

Marclie committed:
"Update linkage class to regenerate vectorized vertices

The code introduces a new boolean parameter (regenerate) to 'to_vector' method in Linkage class. The purpose of this change is to provide an option for regenerating vectorized vertices during the recursive traversal of the graph. This affects the handling of both left and right linked vertices.

The commit significantly refactors the 'to_vector' method to incorporate this change, by adding conditions to check the status of regenerate and modifying how linked vertices are processed accordingly.

Furthermore, with this update, if the regenerate flag is set to true, the vertices are freshly computed with each call to the function, enabling dynamic handling of graph attributes. This is essential for scenarios where the structure of the graph could change dynamically, thereby necessitating a regeneration of the vectors."

Marclie committed:
Fixed issue with sorting and printing of blocked equation (either by spin or range).

This fix required modifications to the recursive expansion of the linkage.cc class. The most notable change was that the declaration of the righthand side tensors of the Term instances created from a LinkagePtr would set the children to be the `left` and `right` of the linkage. This worked since the operators would be expanded ultimately, but ideally the rhs should be generated recursively, which is now the case.

Marclie committed:
Update copy logic to use deep_copy() instead of 'copy_vert'. copy_vert would not make a true deep copy of nested linkages and would cause unintended side effects. By making this function a member of the Vertex class, we can do a recursive deep copy of the left and right vertices in a linkage.

 Similarly, the 'replace_lines' method was added to the Vertex class to allow 'generic' terms to recursively update the lines of nested linkages.

Marclie committed:
now we enforce an assignment of the first occurrence of a term in an equation.

Marclie committed:
Replace lines would break for sigma-build equations. Added check to see if the line exists in the map.

Marclie committed:
Refactor line comparison and optimize linkage expansion added an 'in_order' in the Line class to prevent copying lines. The function 'operator+' also now calls Linkage with the correct argument. Some unnecessary functions and variables were removed to clean up the code and improve readability.

The earlier way of finding matching vertices has been optimized to skip already matched vertices during the substitution procedure.

Marclie committed:
Removed deep copy of equation for testing substitutions now that the vertex deep copies prevent thread races.

Marclie committed:
Removed making linkages inside of `valid_op`. Unnecessary and expensive. Also added some extra checks inside the main `op` function in `Term::substitute`

Marclie committed:
quick change to line hash. allowing hash collisions is cheaper than concatenation.

Marclie committed:
The LineHash is called many, many times. Writing to a string and concatenating has too much overhead. Since for a unique-enough hash we only really care about the booleans and the first character of the line, it is better to directly manipulate the bits to get the result we need.

Marclie committed:
The LineHash is called many, many times. Writing to a string and concatenating has too much overhead. Since for a unique-enough hash we only really care about the booleans and the first character of the line, it is better to directly manipulate the bits to get the result we need.

Marclie committed:
Removed the `int_lines_` member from the `Linkage` class. It has a high memory overhead when creating and destroying linkages. It is only ever used in one function outside the linkage class. The same information is already buried in `int_lines_`, so now there is a function to get the internal lines dynamically.

Marclie committed:
Removed move semantics for `link`. Compiler optimizes it better, whereas explicitly calling std::move has some overhead.

Marclie committed:
Updated term's commenting method and simplified `expand_permutations` function. Removed support temporarily for merging equations. fixed formatting and recognition of scalar temps.

Marclie committed:
Optimized operations for checking similarity of Vertex and Line objects. Created new hash tables and hashing algorithms to prevent excessive object copies and hash collisions. Replaced functions in `is_compatible` and `valid_op` to use O(n) similarity lookup instead of O(n^2) with hash tables

Marclie committed:
Update variables for consistency and readability in pq_graph

Renamed vector<Line> to line_vector for better readability and consistency across multiple references in pq_graph/src/vertex.cc. Also, switched to string concat instead of replacing entire string in pq_graph/src/vertex.cc. Changed some variable names in pq_graph/include/linkage_set.hpp for clarity and conciseness. Applied similar changes to pq_graph/src/consolidate.cc. Made the pointer casting in pq_graph/include/linkage.h use static_pointer_cast instead of dynamic_pointer_cast for performance improvement. Changed the variable type to ConstVertexPtr in pq_graph/src/linkage.cc for consistency.

Marclie committed:
Refactor functions and improve permutation clarity

term_perm_helper.cc was significantly refactored for readability and now handles permuting terms with linkages consistently.

The `to_vector` functions was removed entirely since the recursive process of getting the vertices was a bottleneck. the list of vertices are now built during the construction of the linkage and are saved as a data-member. the new function name to get these vertices is `get_vertices`.

Marclie committed:
functionality for thread-safety was added with the use of std::mutex in the linkage_set class as well as inserting elements into sets in a thread-safe manner.

The changes also improve performance by converting loops to use the more efficient iterator-based approach and by handling resources more judiciously - for example, employing 'reserve' on string objects to avoid unnecessary reallocations.

better handling of empty elements in 'vertex' class and correcting the logic for them in `linkage` class

added functionality to provide a maximum number of contractions within a tmp. However, this is giving strange behavior that is currently being debugged.

Marclie committed:
functionality for thread-safety was added with the use of std::mutex in the linkage_set class as well as inserting elements into sets in a thread-safe manner.

The changes also improve performance by converting loops to use the more efficient iterator-based approach and by handling resources more judiciously - for example, employing 'reserve' on string objects to avoid unnecessary reallocations.

better handling of empty elements in 'vertex' class and correcting the logic for them in `linkage` class

added functionality to provide a maximum number of contractions within a tmp. However, this is giving strange behavior that is currently being debugged.

Marclie committed:
functionality for thread-safety was added with the use of std::mutex in the linkage_set class as well as inserting elements into sets in a thread-safe manner.

The changes also improve performance by converting loops to use the more efficient iterator-based approach and by handling resources more judiciously - for example, employing 'reserve' on string objects to avoid unnecessary reallocations.

better handling of empty elements in 'vertex' class and correcting the logic for them in `linkage` class

added functionality to provide a maximum number of contractions within a tmp. However, this is giving strange behavior that is currently being debugged.

Marclie committed:
Remove redundant copy of equations in consolidate.cc; Refactor linkage comparison

Marclie committed:
Removed redundant `eq_flop_map` in `Equation::test_substitute` and directly use the test_flop_map.

Marclie committed:
Additional logic for keeping equivalent linkages.

Marclie committed:
Additional logic for keeping linkage equivalency

Marclie committed:
This will not be a useful commit message because things are broken. I jsut need to sync with the repo at the moment.

Marclie committed:
Needed to redefine the overload of the vertex operator. Also fixed some formatting for printing with einsum.

Marclie committed:
Refactored the 'shape' class in the by splitting 'o_' and 'v_' into 'oa_','ob_','va_', and 'vb_'. Not sure why I did this in the first place to be honest.

Marclie committed:
Refactor code for more efficient vertex and line comparisons

The linkage vertices are now efficiently stored as a vector instead of an unordered set, improving the search and comparison operations. The scaling of operations is now also handled from a central location, eliminating the need to calculate it multiple times. Hashing now includes the consideration of label permutations. Removed redundant tracking of variables which were replaced with on-demand calculations.

Marclie committed:
improved calculation of the term scaling by explicity adding the scaling of the rhs for the number of permutations. This caused issues when expanding the terms into their permutations in the substitution process.

changed the compute scaling function to return copies of the scaling maps instead of modifying them in-place when given a vector of vertices

Marclie committed:
Reverted back to the commit from 12/1 and readded some changes from recent commit. The data structure I was using before seemed to eliminate candidate contractions from being considered and no nested intermediates would be found. This previous commit was the one that worked the best.

Marclie committed:
Reimplemented the algorithm for line matching after reverting the last commit.

Marclie committed:
Reimplemented the algorithm for line matching after reverting the last commit.

Marclie committed:
Reimplemented the algorithm for line matching after reverting the last commit.

Marclie committed:
Additional check of linkage equivalency just to be safe.

Marclie committed:
changed insertion order for cleaner output.

Marclie committed:
Added check for valid keys for 'max_shape_map'. Allow for expanding the permutations after substitution.
Moved 'make_test_set' to 'consolidate.cc' and changed its functionality to add smaller depth contractions first.
using 'static_pointer_cast' instead of 'dynamic_pointer_cast' for faster conversion of Vertex and Linkage. At some point I should replace this with an explicit cast operator.
Added a faster hashing algorithm for vertices.
Optimized list memory allocation in 'link_and_scale'.

Marclie committed:
Added condition for temps to substitute each other only if the id is greater than the tmp being substituted. This makes it easier to nest temps together.

Marclie committed:
Added condition for temps to substitute each other only if the id is greater than the tmp being substituted. This makes it easier to nest temps together.

Marclie committed:
Added condition for temps to substitute each other only if the id is greater than the tmp being substituted. This makes it easier to nest temps together.

Marclie committed:
changed line hash

Marclie committed:
"permute_eri_" was addded to options to control whether eris should be permuted to a common order to help with substitution.
Also, the "t1_transform" option was removed since this can be controlled on the pdaggerq side by not including t1 in the cluster amplitudes. Readded the "merge_terms()" function, but set it up to use a flag that defaults to false.

The python printing function in the Term class needed a crucial addition to prevent einsum from making shallow copies by multipying by a number. Also suppressed the printing of amplitudes as a map.

Marclie committed:
Add print override to Term class for custom printing

Marclie committed:
Add print override to Term class for custom printing

Marclie committed:
Changes in formatting of the printed name. Now instead of using the tensor type to determine if it should be printed as a map, there is a manual boolean `format_map_` to toggle it on and off. Now the assignment vertex is off by default.

Marclie committed:
troubleshooting

Marclie committed:
The `Vertex::same_lines()` function was inlined but used in a static function within Term. This did not fail on my local build, but did fail elsewhere.

Marclie committed:
Removed throwing error for mismatching blkstrings and rank. This would fail anyways and does not work when doing eom builds

Marclie committed:
Give user a warning when there are more than 10000 possible contractions. Now prints out the progress during optimization

Marclie committed:
Changed printing ratio for more updates and also fixed an issue when dividing by zero to find the common coefficient.

Marclie committed:
Changed printing ratio for more updates and also fixed an issue when dividing by zero to find the common coefficient.

Marclie committed:
Changed printing ratio for more updates and also fixed an issue when dividing by zero to find the common coefficient.

Marclie committed:
Okay, print_ratio can be zero no longer.

Marclie committed:
Block prints out weird when returning a char. now returns a string.

Marclie committed:
printing has been made cleaner

Marclie committed:
printing has been made cleaner

Marclie committed:
The commit introduces changes that improve the logic for the sigma-vector build and reintroduce the merge algorithm. The changes aim to enhance code functionality and readability, with amendments to several files, including 'pq_graph.h', 'equation.cc', 'pq_graph.cc', and 'term_substitute.cc'. Iterations over all linkages have been optimized for better performance, and the process to collect new scaling after a successful substitution has been streamlined. In addition, unnecessary code fragments were removed, and more comprehensive comments were added for clarity. The code related to the merge algorithm has been refactored to be more efficient and easy to understand. The 'allow_merge' condition now accounts for whether the 'format_sigma' flag is active to control the merging of terms.

Marclie committed:
Full recursion when sorting temps is not optimal for freeing memory. Only recurse at a depth of 1

Marclie committed:
Print function needed adjustments

Marclie committed:
Print function needed adjustments

Marclie committed:
Print function needed adjustments

Marclie committed:
additional logic for merging temps

Marclie committed:
additional logic for merging temps

Marclie committed:
additional logic for removing candidate temps

Marclie committed:
It is better not to assume no intermediate when substituting intermediates since that gives an inconsistent scaling comparison when expanding permutations.

Marclie committed:
substituted vertices in term_substitute.cc are now re-ordered instead of re-computing the scaling.

now using a multimap instead of a vector of pairs relating flop scaling maps to linkages. Connections between lines in linkage.h are now stored in a vector of arrays instead of a vector of pairs for better performance. Instead of inserting the connections in order, we change the equality operator to check if all pairs from one linkage exists in the other.

Marclie committed:
Reset data structures in consolidate.cc and linkage.cc for performance.

consolidate.cc did not support multithreading with changes. linkage.cc had issues with malloc.

Marclie committed:
The algorithm of batching substitutions has been reworked.

 Previously, only the best substitution was found and replaced once per iteration. Now, if enabled, we compute the scaling change from every substitution, sort them by cost, and substitute them in ascending cost. Ideally, the best substitutions should prevent worse ones from being substituted. Even in that case, the code supports nested contractions that would circumvent this.

Marclie committed:
condensed the int_connec_, l_ext_idx_, and r_ext_idx_ vectors into a single vector, connec_map_. Updated the hashing algorithms for both vertices and linkages to have less collisions.

Marclie committed:
Added `batch_size_` to the options and adjusted `max_temps_`, `max_depth_`, `max_shape_` to handle longs instead of ints. Fixed the `remove_redundant_tmps()` method to correctly remove single term intermediates. `substitute()` method was revised for optimal batching. Other minor changes for readability and efficiency.

Marclie committed:
fixed printing for c++ code.

Marclie committed:
It is still necessary to insert lines in order for the linkage. Using a traditional map is not sufficient.

Marclie committed:
The `remove_redundant_tmps` had a mistake that did not correctly identify and substitute all temp that needed to be removed within nested linkages. This required recursive logic to traverse sequences of nested linkages and remove the target ones.

Marclie committed:
The `remove_redundant_tmps` is more consistent when done at the printing stage.

Marclie committed:
No need to reapply substitutions if we are removing redundant linkages.

Marclie committed:
Changed nthreads to allow for max threads with negative input

Marclie committed:
Changed sort function to treat different examples properly.

Marclie committed:
Changed sort function to treat different examples properly.

Marclie committed:
Give up on removing redundant temps until I KNOW it works for all examples.

Marclie committed:
remove 'np.' from printing. It is unnecessary when using `from numpy import einsum`

Marclie committed:
added an example script to run

Marclie committed:
adjusted method for tracking the scaling for different parts of the optimization by expanding permutations before anything

Marclie committed:
Changed how external lines are populated to ensure a consistent ordering. updating directly from the line populations does not distinguish whether the line came from the right or the left. It is better to keep track of the external indicies in a boolean array and reiterate over the left and right lines. This change allows for the removal of redundant temps to work (however it can only be called once for printing)

Marclie committed:
Added numerical test for ccsdt using the parser python output

Marclie committed:
Merge remote-tracking branch 'origin/pq_graph' into pq_graph

Marclie committed:
refactored the link function to have more descriptive names and to better allocate memory for the lines to prevent over allocating memory.

There was a typo for regenerating the internal lines of a linkage that has been fixed.

Marclie committed:
scalars do not need to be updated since they require little storage.

Marclie committed:
the expanded permutations should not be reordered by default. that should be handled outside of this function. However, the scaling does need to be updated.

I've also fixed the flag that indicates if equations should be formatted for a sigma build or not.

Marclie committed:
Added functionality to the `add` function of pq_graph. Now an additional optional argument can be provided to indicate a desired order for the indices of the assignment vertex. Previously if two equations with the same name were added to the pq_graph object, the second call to add would overwrite the equations from the first call. Instead a warning is printed and the terms of the second call are added to the equation created from the first call.

Marclie committed:
adding more numerical tests and equation generation.

Marclie committed:
added logic to make the sigma lines always appear first.

Marclie committed:
Updated examples

Marclie committed:
Needed to account for edge case when the term has no operators, but a nonzero coefficient.

Marclie committed:
reimplemented the remove redundant function since it was not comprehensive.

Marclie committed:
line information was not correctly transferred when creating delta functions.

Marclie committed:
assignment of new terms missing

Marclie committed:
Fixed the implementation of removing unused intermediates for robustness.

Marclie committed:
Updated example

Marclie committed:
Changed procedure for extracting out scalars since it was not robust enough for extracting them all. This affects the `consolidate.cc`, `equation.cc`, and `term.cc` classes.

Some refactoring has also been added by replacing `Vertex::deep_ptr_copy` with `Vertex::clone_ptr`. the last argument in `Term::operate_subsets` was given a more descriptive name as `terminate_op` to more clearly communicate the purpose of the callback function

Marclie committed:
The coefficients would not print out consistently for small numbers, so I changed the procedure to print the float as a string and count the significant digits using a loop.

Marclie committed:
modified some flags for formatting of dot file

Marclie committed:
updated sorting function for brevity

Marclie committed:
implemented density fitting for terms by removal of eri.

Marclie committed:
replaced hash map with sorted lists to determine eligible substitutions

Marclie committed:
added density fitting support, quick fix for max_sizes, and now declarations of tmps are handled externally and not by the sort function.

Marclie committed:
use std::include instead of building a set

Marclie committed:
updated hashing algorithm slightly

Marclie committed:
attempted to implement a comparison to see if a linkage is equivalent up to a permutation. Not successful, but close

Marclie committed:
updated examples with more recent version of parser

Marclie committed:
Fixed procedure for adding declarations such that they are added in order of use.

Marclie committed:
Merge branch 'master' into pq_graph

Marclie committed:
supress pybind deprecation warnings

Marclie committed:
more comments for pq

Marclie committed:
The addition of `set(Python3_FIND_STRATEGY LOCATION)` changes the `FindPython.cmake` procedure for finding the python executable. This procedure prevents CMake from using the system python interpreter, and uses the python that exists in the current shell environment (i.e. a conda environment).

Marclie committed:
REQUIRED is not a language

Marclie committed:
Optimize Linkage class:
    Add tree balancing, cloning, and equality comparison
    Rename functions for clarity
Enhance Equation class:
    Store equation linkage, add term sorting and condition tracking
    Optimize linkage generation and term merging
Update Term and PQGraph:
    Add condition tracking, deep cloning, and rename functions
    Optimize substitution and parallelization
    Update defaults and naming for clarity

Marclie committed:
typo for sorting reuse tmps. I was supposed to use 'r' not 't' ....

Marclie committed:
Update with all recent changes that improves condition determination; improves scalar substitution with flag to remove them; adds example files that uses the parser output; improves initialization of Terms from pq_helper. other changes.

Marclie committed:
update readme and logo

Marclie committed:
update readme and logo

Marclie committed:
update readme and logo

Marclie committed:
Merge branch 'refs/heads/master' into pq_graph

# Conflicts:
#	CMakeLists.txt
#	pdaggerq/pq_helper.cc

Marclie committed:
Readd serialize function

Marclie committed:
Update README.md

Marclie committed:
Update README.md

Marclie committed:
Update README.md

edeprince3 commented:
some openmp fixes in cmake are necessary. 

on my macbook, cmake didn't notice that i didn't have openmp libraries installed. once they are installed, there are still issues with finding the right compilers.

on my home linux box, the package compiles, but there is some linking problem. i get runtime errors about missing omp symbols

Marclie committed:
Added ifdefs to make openmp optional.

Marclie commented:
I added the omp fixes. Please let me know if that fixes the linking issues on your end. Thank you!

Marclie committed:
Added ifdefs to make openmp optional. Disabled interprocedural optimization.

Marclie committed:
Added ifdefs to make openmp optional. Disabled interprocedural optimization.

Marclie committed:
added sstream include for clang compiler (not included in string.h like in GNU)

Marclie committed:
add setting to fix generation of the DOT file. Needed to explicitly create the assignment vertex for scalars and tmps

Marclie committed:
initialize values for time

Marclie committed:
update README.md

Marclie committed:
add explicit overrides to prevent warnings

Marclie committed:
add explicit overrides to prevent warnings

Marclie committed:
Merge branch 'refs/heads/master' into pq_graph

Marclie committed:
Improved logic to identify left and right sigma operators and keep the indices in the correct order.

Marclie committed:
Changed reorder logic to keep permutation that both minimizes flop scaling and preserve a contraction order that is closest to the lhs vertex.

Marclie committed:
Changed reorder logic to keep permutation that both minimizes flop scaling and preserve a contraction order that is closest to the lhs vertex.

Marclie committed:
Fixed reorder logic to keep permutation that preserves best index order for the contraction.

Marclie committed:
Fixed reorder logic to keep permutation that preserves best index order for the contraction.

Marclie committed:
moved lambda function to count permutations of indicies to member function of the Term class. Also incorporated the logic into the substitution instead of just the reordering.

Marclie committed:
renamed `print_trial_index` to `use_trial_index`

Marclie committed:
Changed implementation of tree sort, which was essentially a no-op the way it was implemented.

Marclie committed:
Allow for reordering during substitution.

Marclie committed:
Do not use the incremental method anymore for substitutions.

Marclie committed:
adjusted logic of substitute to account for different storage costs.

Marclie committed:
logic for binary linkage of vertices is unnecessary when using tree sort

Marclie committed:
dumb typo for comparison.

Marclie committed:
added easy printing function for `scaling_map.hpp`

Marclie committed:
added convenient constructor to make a scaling map from vector of scales

Marclie committed:
fixed printing of analysis at the end. inconsistent when not optimizing.

Marclie committed:
Changed implementation of tree sort.

Marclie committed:
changed default for extracting intermediates with a sigma build to false.

Marclie committed:
update ccsdt.py example

Marclie committed:
tree sort algorithm seems to make no difference to final output. removing for now.

Marclie committed:
removing empty vertices from output of the `link_vector()` function.

Marclie committed:
Added total timer to class scope.

Marclie committed:
Changed hashing function for a linkage

Marclie committed:
did not add omp flags to CXX flags.

Marclie committed:
modified batching algorithm to determine whether linkages should be formed incremenetally.

Marclie committed:
include c flags for openmp

Marclie committed:
updated batching again.

Marclie committed:
oops, forgot to comment back in the pragma.

Marclie committed:
allow printing better printing for multiple substitution calls

Marclie committed:
subset conditions for valid linkages would eiminate nested intermediates as candidates.

Marclie committed:
fixed printing for removing an intermediate.

Marclie committed:
undo printing change.

Marclie committed:
No longer using tree sort, so we are sorting binary ops by name again to prevent redundancy.

Marclie committed:
Refactored `safe_clone()` into `shallow` because of misleading naming after functionality changes
Changed hashing algorithm for linkages

Major problem with merging terms together after recent commits. The `term_linkage()` function did not correctly update the term linkage when called. Now it explicitly computes the term linkage until the cause is identified.

Marclie committed:
changed algorithm for merging terms together with an 'easier' to read hashmap

Marclie committed:
updated full code ccsdt example with pq_graph

Marclie committed:
Had to account for less than two vertex terms when merging

Marclie committed:
changed name for dot file

Marclie committed:
cleaner generation of DOT file for graph visualization.

Marclie committed:
Added comments to visualizer and removed static variables.

Marclie committed:
Whoops. This is a huge mistake... a miracle anything even worked before... must be because I overload this operator for linkages and rarely gets called.

Marclie committed:
Change naming scheme for sigma operators.

Marclie committed:
adding functionality for fusion; still experimental so it will be disabled. Also added multiple tests for the qed codes.

Marclie committed:
making the `merge_intermediates` function a no-op for now

Marclie committed:
refactored tests to make them easier to comment in and out.

Marclie committed:
Merge branch 'refs/heads/master' into pq_graph

# Conflicts:
#	examples/qed_ccsd.py
#	examples/qed_eom_ccsd.py
#	test/pq_test.py
#	test/reference_outputs/qed_ccsd.ref
#	test/reference_outputs/qed_eom_ccsd.ref

Marclie committed:
Merge branch 'refs/heads/master' into pq_graph

# Conflicts:
#	pdaggerq/pq_helper.cc
#	pdaggerq/pq_string.cc

Marclie committed:
Changed designation of temp ids to belong to the corresponding temp types.

Marclie committed:
miscellaneous information of linkage was not copied when the linkage was not explicitly cloned. May not be necessary since the copy constructor is called, but it doesn't hurt.

Marclie committed:
Scalars were not properly being substituted. Ids would be inconsistent. Fixed by adding checks that the substituted linkages go to the correct linkage type.

Marclie committed:
Nested scalars now work and can be checked here.

Marclie committed:
More explicit call to copy constructor.

Marclie committed:
Merge remote-tracking branch 'refs/remotes/origin/pq_more_photons' into pq_graph

# Conflicts:
#	pdaggerq/pq_tensor.h

Marclie committed:
Adding support for more than one photon

Marclie committed:
Remade CCSDT example

Marclie committed:
Added copy constructors and assignment operators. Shared pointers were not deep copied as expected.
move constructors are set to default.
made a function on the python side to clone the pq object.

Marclie committed:
Changed printing to add an extra space. Makes it easier to copy and paste code blocks since autoformatters in IDEs do not like starting lines with '//' for some reason.

Marclie committed:
Fixed generation of DOT file when term has no rhs

Marclie committed:
Fixed generation of DOT file with temps

Marclie committed:
updated use of '^' with '=='. Easier readability.

Marclie committed:
updated use of '^' with '=='. Easier readability.

Marclie committed:
added visualization for self contractions.

Marclie committed:
Revert "updated use of '^' with '=='. Easier readability."

This reverts commit 1893970348ce8450a188abf4cd01e07d6b8b1e92.

Marclie committed:
Revert "updated use of '^' with '=='. Easier readability."

This reverts commit 09f6ac758339bcb665c8aba45f4a43d0c868ba1c.

Marclie committed:
Oops, meant to replace '^' with '!='

Marclie committed:
fixed significant operator treatment

Marclie committed:
removed subgraphs for temps (made the viz look ugly)

Marclie committed:
faulty addition for alpha virtuals. Shouldn't change results.
Added parameters to keep track of total alpha/beta and occ/vir.

Marclie committed:
added check for same_a_b for separating equations only needed for different alpha/beta

Marclie committed:
removed explicit constructor. No other object should automatically convert into Line objects. also changed initialization.

Marclie committed:
removed the check_a_b functionality (not useful)

Marclie committed:
Updated QED scripts to properly clone the pq object.

Marclie committed:
Merge branch 'refs/heads/master' into pq_graph

# Conflicts:
#	examples/qed_eom_ccsd.py
#	pdaggerq/pq_helper.cc

Marclie committed:
Add fence for C++

Marclie committed:
include stl containers with pybind again (whoops)

Marclie committed:
Just use pybind to manage getting the correct python

Marclie committed:
fixed fencing to appear after the destructor call

Marclie committed:
removed TODO since it was done

Marclie committed:
Had to change import order for pq_graph
